Implement exp2

ShawnZhong opened the pull request:
X-ref https://github.com/pytorch/pytorch/issues/38349

mruberry commented on torch/_torch_docs.py:
```diff
@@ -2086,6 +2086,26 @@ def merge_dicts(*dicts):
+""" + r"""
+Args:
+    {input}
+    {out}
```
See how torch.add handles its kwarg out with 
```
Keyword arguments:
    {out} 
```

mruberry commented:
@ailzhang I think the XLA failure here is misleading, and XLA just doesn't understand how to lower this function?

mruberry reviewed (changes requested):
This is an excellent, thorough PR, thank you @ShawnZhong! I've asked @ailzhang to comment on the XLA issue, and I made a note on a very modest documentation change. One thing that appears to be missing is a derivative, however. The exp derivative is here:

https://github.com/pytorch/pytorch/blob/a25062ab50bf7633a94c56d3fd11c00f4c5fb8f6/tools/autograd/derivatives.yaml#L346

And you can add autograd tests by mimicking the exp registration here: 

https://github.com/pytorch/pytorch/blob/a25062ab50bf7633a94c56d3fd11c00f4c5fb8f6/torch/testing/_internal/common_methods_invocations.py#L223

After these three issues are addressed I think we're good to go!

ShawnZhong committed:
Implement exp2

Update _torch_docs.py

Update test_torch.py

debug

debug

Update

Revert "debug"

This reverts commit 4cc22744a6bbce473c62a3cd80ff9e43c1606199.

Revert "debug"

This reverts commit fdcff62d013f6c0fe66d106cdcdb8a0b98c873fc.

Update

bugfix

Update

remove builtins

bug fix

mruberry commented on test/test_torch.py:
```diff
@@ -18633,6 +18638,8 @@ def inner(self, device, dtype):
     ('exp', '', _small_3d, lambda t, d: [], 1e-2, 1e-2, 1e-5, _float_types),
     ('exp', 'samll', lambda t, d: _small_3d(t, d).clamp(-1, 1),
         lambda t, d: [], 1e-2, 1e-2, 1e-5, _float_types, [torch.bfloat16]),
+    ('exp2', 'samll', lambda t, d: _small_3d(t, d).clamp(-1, 1),
```
"samll" -> "small", would you correct this in the preceeding exp test case, too?

Also, you could add another test cases analogous to the first exp test case:

```
('exp2', '', _small_3d, lambda t, d: [], 1e-2, 1e-2, 1e-5, _float_types),
```

mruberry commented on tools/autograd/derivatives.yaml:
```diff
@@ -367,6 +367,9 @@
   self: grad * result
 
+- name: exp2(Tensor self) -> Tensor
+  self: grad * result * 0.6931471805599453
```
@albanD Are you happy with this approach or would you prefer the approach taken in, for example, the deg2rad PR (https://github.com/pytorch/pytorch/pull/38852/files)? 

@ShawnZhong, did you chose this version of ln(2) because it goes to the first decimal s.t. the number is no longer exactly representable as a double?

albanD replied:
I would be slightly concerned about the precision here for the double case. Is that done on purpose?
If we want full precision, adding a constexpr sounds like a good way to easily define the full precision value.

mruberry replied:
IEEE 1003.1 says that math.h (AKA cmath) should define M_LN2, which we should try to use instead of our own macro/constexpr expression.

mruberry reviewed (changes requested):
Hey @ShawnZhong! Thanks for the thorough PR. Overall things look very good. I made a few comments, but a bigger issue is that this PR is touching several largely separate components of PyTorch.

Would you separate this PR into this PR minus the JIT components, then submit two separate followup PRs: one for the tensor expressions fuser and another for the codegen/cuda fuser. That will let those teams have a chance to review your work.

mruberry commented on docs/source/amp.rst:
```diff
@@ -135,6 +135,7 @@ Ops that can autocast to ``float32``
 ``dist``,
 ``erfinv``,
 ``exp``,
+``exp2``,
```
Is this doc update correct? I don't think you're changing how AMP works, right?
